src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,412 @@
+-- result of adding a larger-than-page value to a run, which made the page
+-- contents exceed the size of a single disk page.
+--
+-- The string of bytes is padded to the target disk-page size (4K).
+--
+-- This builder could be used to serialise into a strict bytestring, lazy
+-- bytestring, or into existing buffers.
+pageBuilder :: PageAcc -> BB.Builder
+pageBuilder PageAcc{..} =
+    -- (1) directory of components
+       BB.word16LE pageSizeNumElems
+    <> BB.word16LE pageSizeNumBlobs
+    <> BB.word16LE offKeyOffsets
+    <> BB.word16LE 0
+    -- (2) an array of 1-bit blob reference indicators
+    <> dfoldMap BB.word64LE (bmbits pageBlobRefBitmap)
+    -- (3) an array of 2-bit operation types
+    <> dfoldMap BB.word64LE (cmbits pageOperations)
+    -- (4) a pair of arrays of blob references
+    <> dfoldMap BB.word64LE pageBlobRefOffsets
+    <> dfoldMap BB.word32LE pageBlobRefSizes
+    -- (5) an array of key offsets
+    <> dfoldMap BB.word16LE pageKeyOffsets
+    -- (6) an array of value offsets
+    <> case pageValueOffsets of
+          Left   offsets -> dfoldMap BB.word16LE offsets
+          Right (offset1, offset2) -> BB.word16LE offset1
+                                  <> BB.word32LE offset2
+    -- (7) the concatenation of all keys
+    <> dfoldMap rawKey (unStricterList pageKeys)
+    -- (8) the concatenation of all values
+    <> dfoldMap rawValue (unStricterList pageValues)
+    -- padding
+    <> fold (replicate (fromIntegral paddingBytes) (BB.word8 0))
+  where
+    dfoldMap f = getDual . foldMap (Dual . f)
+
+    PageSize{..} = pageSize
+
+    n = pageSizeNumElems
+    b = pageSizeNumBlobs
+
+    -- Offset to component (5)
+    offKeyOffsets =
+         8                                         -- size of (1)
+      + (n + 63) `shiftR` 3 .&. complement 0x7     -- size of (2)
+      + (2 * n + 63) `shiftR` 3 .&. complement 0x7 -- size of (3)
+      + (4 + 8) * b                                -- size of (4)
+
+    -- Offset to component (7)
+    offKeys =
+        offKeyOffsets
+      + 2 * n                             -- size of (5)
+      + (if n == 1 then 6 else 2 * (n+1)) -- size of (6)
+
+    -- Thes values start right after the keys,
+    (pageKeyOffsets, offValues) =
+        case scanr (\k o -> o + sizeofKey16 k) offKeys (unStricterList pageKeys) of
+          []         -> error "impossible"
+          (vo : kos) -> (kos, vo)
+
+    pageValueOffsets = case unStricterList pageValues of
+      [v] -> Right (offValues, fromIntegral offValues + sizeofValue32 v)
+      vs  -> Left (scanr (\v o -> o + sizeofValue16 v) offValues vs)
+
+    rawKey :: SerialisedKey -> BB.Builder
+    rawKey = BB.shortByteString . toShortByteString
+
+    rawValue :: RawValue -> BB.Builder
+    rawValue EmptyRawValue                  = mempty
+    rawValue (RawValue i j (ByteArray ba#)) = shortByteStringFromTo i j (SBS ba#)
+
+    paddingBytes :: Word64
+    paddingBytes = 4096 - pageSizeNumBytes `rem` 4096
```

Comment:
Does this account for larger-than-page values, where this could get negative?

Reply:
Note that `rem` has higher precedence than `(-)`, so it can't go negative

Reply:
Ah, right. But then we get the issue that ``4096 `rem` 4096`` is 0, so after an exactly full page we add another page of padding?

I also ended up with something like ``4096 - ((size - 1) `mod` 4096 + 1)`` in my PR and this kind of calculation shows up again and again. Maybe we should have a utility function somewhere?